Standardizing Fulcio Certificate Extensions (#945)
sigstore/fulcio: docs/oid-info.md

Assigned Numbers Authority to help identify and organize additional metadata in
code signing certificates issued by Fulcio instances. This document aims to
provide a simple directory of values in use with an explanation of their
meaning.

## Requirements to support signing with CI/CD workload identities

In order to support Sigstore code signing with CI/CD based workfload identities the following claims must be included in the OIDC ID Token. See example claim values for each extension in the detailed [Directory](#directory). 

Requirements:

- MUST include `iss` claim for `Issuer` extension.
- MUST include claim to support: `Build Signer URI` that identifies the specific build instructions that are responsible for signing.
- MUST include claim to support: `Build Signer Digest` which is an immutable reference to a specific version of the build instructions that are responsible for signing.
- MUST include claim to support: `Runner Environment` that differentiates between builds that took place in platform-hosted cloud infrastructure or customer-hosted infrastructure.

Recommended:

> - SHOULD include `iss` that uniquely identifies ID tokens originating from the CI/CD system, e.g. not shared with OIDC OAuth 2.0 tokens for email/username logins.

Reviewer: What is the reason for this requirement?

Reviewer: Makes sense, although you could still run into issues if the OIDC issuers are run within the same trust domain such that if one is compromised, both could be. For example a CI issuer issuing a token for an email address, or an email-based issuer issuing a token for a CI identity (we do add some checks in fulcio to make sure a subject looks like an email when it's expected, I would hesitate to always enforce that a CI identity is not an email though)

- SHOULD include claim to support: `Source Repository URI`
- SHOULD include claim to support: `Source Repository Digest`
- SHOULD include claim to support: `Source Repository Ref`
- SHOULD include claim to support: `Source Repository Identifier`
- SHOULD include claim to support: `Source Repository Owner URI`
- SHOULD include claim to support: `Source Repository Owner Identifier`

Nice-to-haves:

- MAY include claim to support: `Build Config URI`
- MAY include claim to support: `Build Config Digest`
- MAY include claim to support: `Build Trigger`
- MAY include claim to support: `Run Invocation URI`

## Terminology

- `Identifier`: Immutable resource identifier, e.g. uuid/primary key ID
- `URI`: SHOULD be a fully qualified URL when available. MAY be a mutable resource identifier, e.g. `https://scm.com/example/repository`
  - Fully qualified URL: Complete URL with protocol.

> - `Digest`: Immutable source repository reference, e.g. git commit SHA

Reviewer: nit: I assume a digest isn't always a source repository ref, like for "build config digest", so should this instead say "Output of a cryptographic hash function"

## Directory

Note that all values begin from the root OID 1.3.6.1.4.1.57264 [registered by Sigstore][oid-link].

When adding additional OIDs under the root, please update the above link with the child OID.

> GitHub Workflow specific OID extensions have been deprecated in favour of provider generic extensions starting from 1.3.6.1.4.1.57264.1.8.

Reviewer: nit: "favour" -> "favor" (I like UK English spellings, but just being consistent)

## Description

Sigstore maintains its own Private Enterprise Number ([57264](http://oid-info.com/get/1.3.6.1.4.1.57264)) with the Internet
Assigned Numbers Authority to help identify and organize additional metadata in
code signing certificates issued by Fulcio instances. This document aims to
provide a simple directory of values in use with an explanation of their
meaning.

## Requirements to support signing with CI/CD workload identities

In order to support Sigstore code signing with CI/CD based workfload identities the following claims must be included in the OIDC ID Token. See example claim values for each extension in the detailed [Directory](#directory). 

Providers MAY choose to emit extension value in other formats to generic examples, and consumers MUST NOT assume the generic example format.

Requirements:

- MUST include `iss` claim for `Issuer` extension.
- MUST include claim to support: `Build Signer URI` that identifies the specific build instructions that are responsible for signing.

> - MUST include claim to support: `Build Signer Digest` which is an immutable reference to a specific version of the build instructions that are responsible for signing.

Reviewer 2: How should these attributes be supported by services like Buildkite which offer dynamic pipelines, i.e. build instructions are not immutable and change over the course of an execution? I suppose we could include a reference to the _initial_ build instructions?

Reviewer 2: Ah, I see, the directory examples suggest this could be a link to the source repository at a specific commit sha. That doesn't represent the whole signing environment for us, the instructions also involve the running pipeline, but it might be close enough.

Reviewer 3: Yes, so ideally this would be an immutable reference to the Buildkite pipeline steps at the time the build ran.

How are Buildkite pipeline steps stored / versioned? If I update the pipeline steps **after** a build has run, is there a URL I can go to that references the previous version of the pipeline steps (e.g. the pipeline steps when the build ran)?

Requirements:

- MUST include `iss` claim for `Issuer` extension.
- MUST include `iss` that uniquely identifies ID tokens originating from the CI/CD system, e.g. not shared with OIDC OAuth 2.0 tokens for email/username logins.
- MUST include claim to support: `Runner Environment` that differentiates between builds that took place in platform-hosted cloud infrastructure or customer-hosted infrastructure.
- MUST include claim to support: `Build Signer URI` that identifies the specific build instructions that are responsible for signing.
- MUST include claim to support: `Build Signer Digest` which is an immutable reference to a specific version of the build instructions that are responsible for signing.

Recommended:

- SHOULD include claim to support: `Source Repository URI`
- SHOULD include claim to support: `Source Repository Digest`
- SHOULD include claim to support: `Source Repository Ref`
- SHOULD include claim to support: `Source Repository Identifier`
- SHOULD include claim to support: `Source Repository Owner URI`
- SHOULD include claim to support: `Source Repository Owner Identifier`

Nice-to-haves:

> - SHOULD include claim to support: `Build Config URI`
> - SHOULD include claim to support: `Build Config Digest`
> - SHOULD include claim to support: `Build Trigger`
> - SHOULD include claim to support: `Run Invocation URI`

Author: ```suggestion
- MAY include claim to support: `Build Config URI`
- MAY include claim to support: `Build Config Digest`
- MAY include claim to support: `Build Trigger`
- MAY include claim to support: `Run Invocation URI`
```

## Terminology

- `Identifier`: Immutable resource identifier, e.g. uuid/primary key ID

> - `URI`: SHOULD be a fully qualified URL when available. CAN be a mutable resource identifier, e.g. `https://github.com/sigstore/fulcio`

Reviewer: nit, can we use MAY rather than CAN (from https://www.rfc-editor.org/rfc/rfc2119)

This contains the `issuer` claim from the OIDC Identity Token that was
presented at the time the code signing certificate was requested to be created.
This claim is the URI of the OIDC Identity Provider that digitally signed the
identity token. For example: https://token.actions.githubusercontent.com.

> ### 1.3.6.1.4.1.57264.3.1 | Source URI

Reviewer 4: The OID prefix `1.3.6.1.4.1.57264.1` is specifically registered for Fulcio so I think its important that these new definitions all use that prefix instead of changing to the `1.3.6.1.4.1.57264.3` prefix. Ref: http://oid-info.com/get/1.3.6.1.4.1.57264.1

Reviewer 4: Ah TIL! I agree that changing the prefix seems unnecessary and can lead to confusion though.

Author: Yeah I can see how changing the prefix is confusing unless we change all of them. Can the suffix increment to double digits? e.g. `1.3.6.1.4.1.57264.1.12`, assuming we want to keep all existing extensions as they and add the new ones following `1.3.6.1.4.1.57264.1.7 | OtherName SAN`. So "Source URI" would end up as `1.3.6.1.4.1.57264.1.8` etc.

Reviewer: yep, the suffix can increase to two digits. I think that would be fine to start from `.8`.

Should include a fully qualified source repository URL that the build was based
on.

### 1.3.6.1.4.1.57264.3.2 | Source Digest

An immutable reference to a specific version of the source code that the build
was based upon. Should include the digest type followed by the digest, e.g.
`sha1:abc123`.

### 1.3.6.1.4.1.57264.3.3 | Source Ref

The source ref that the build run was based upon. For example: `refs/head/main`.

> ### 1.3.6.1.4.1.57264.3.4 | Build Config URI

Reviewer 5: I have a question about what to do with external modules. Let's take GHA as an example, one can use an external action and reference by a branch and not an presumably immutable SHA or tag:

```yaml
uses: publisher-username/repo-name@branch-name
```

In that case the `branch-name` will be evaluated at the time of build execution. Similar example is possible with CircleCI Orbs and with other CI systems.

I wonder if there is a need to have URIs to all the config files that affected the build. Or should this transitive nature be reflected in the following "Build Config Digest"?

Author: Interesting point, I don't think we have any good way of capturing what exact commit external actions where pulled from in GitHub Actions but this would be nice to know for sure. 

My thinking here was that `Build Config` would only refer to the initiating or top-level config (or entry point in SLSA lingo) that was used when running the job in response to the "Build Trigger". The digest would be the commit sha of this top-level config. Given this you could see the exact top-level workflow that was used but not what exact transitive actions where used unless you did some kind of date based commit lookup in each action at the time the job ran?

A reference to the initiating build instructions.

### 1.3.6.1.4.1.57264.3.5 | Build Config Digest

An immutable reference to the specific version of the top-level build
instructions. Should include the digest type followed by the digest, e.g.
`sha1:abc123`.

> ### 1.3.6.1.4.1.57264.3.6 | Build Siger URI

Reviewer 6: ```suggestion
### 1.3.6.1.4.1.57264.3.6 | Build Signer URI
```

Reference to specific build instructions that are responsible for signing. Can be the same as Build Config URI. For example a reusable workflow in GitHub Actions or a Circle CI Orbs.

> ### 1.3.6.1.4.1.57264.3.7 | Build Siger Digest

Reviewer 5: Could you please clarify this a bit similar to Build Siger URI? Is it up for a provider to decide what this is? E.g. SHA of an external GitHub Action or a hash of a bash script that did signing?

Author: Yes the intention here was to provide a immutable reference to the Build Signer URI. So in GitHub Actions case the URI would be the path with branch/tag ref of a reusable workflow. The digest would be the commit sha of the reusable workflow file. In GitHub's case the workflow ref usually contains a mutable branch/tag unless it's been referenced via a commit in the triggering workflow.

Hadn't thought about this directly referencing a bash script. I wonder if this is something that would be possible as all these values will be pulled from the CI/CD providers OIDC claims, so the CI provider would need to extract the location/digest of the script and place it in the token claims.

My intention with calling it the "build signer" was to make it clear that it's the job/container/thing where the CI/CD's OIDC token was requested. It's technically possible to pass this token back to another job where the signing happens though.

In GitHub actions case "Build Signer" and "Build Config" are identical if you're **not** using a reusable workflow. I imagine this would also be the case for Circle CI if using Orbs or not.

Another name idea for "Build Signer" could be "Build Identity" but potentially more confusing.

> ### 1.3.6.1.4.1.57264.3.7 | Build Siger Digest

Reviewer 6: ```suggestion
### 1.3.6.1.4.1.57264.3.7 | Build Signer Digest
```

An immutable reference to the specific version of the build instructions that is responsible for signing. Should include the digest type followed by the digest, e.g. `sha1:abc123`.

### 1.3.6.1.4.1.57264.3.8 | Build Trigger

> The event or action that triggered the build.

Reviewer 6: ```suggestion
The event or action that triggered the build.

```

> ### 1.3.6.1.4.1.57264.3.9 | Runner Environment

Reviewer 5: `platform-hosted` environments are always ephemeral. In case of `self-hosted` environment they are most likely not ephemeral but can be. Should there be a some optional clarifications like version of environment for `platform-hosted` or host name for `self-hosted`?

Author: Good point, we could construct this as a user-agent string that would support versions and extra metadata. 

I'm anticipating that each CI provider probably ends up having custom values that make most sense for them and would be hard to standardize across providers. In GitHub's case, it might be clearer to say `runner_env: github` instead of `runner_env: platform-hosted`.

`sha1:abc123`.

### 1.3.6.1.4.1.57264.1.16 | Build Signer URI

Reference to specific build instructions that are responsible for signing. Can be the same as Build Config URI. For example a reusable workflow in GitHub Actions or a Circle CI Orbs.

Note this value would also be set in the Subject Alternative Name (SAN).

### 1.3.6.1.4.1.57264.1.17 | Build Signer Digest

An immutable reference to the specific version of the build instructions that is responsible for signing. Should include the digest type followed by the digest, e.g. `sha1:abc123`.

### 1.3.6.1.4.1.57264.1.18 | Build Trigger

The event or action that triggered the build.

### 1.3.6.1.4.1.57264.1.19 | Runner Environment

For platforms to specify whether the build took place in platform-hosted cloud infrastructure or customer-hosted infrastructure. For example: `platform-hosted` and `self-hosted`.

## 1.3.6.1.4.1.57264.2 | Policy OID for Sigstore Timestamp Authority

Not used by Fulcio. This specifies the policy OID for the [timestamp authority](https://github.com/sigstore/timestamp-authority)
that Sigstore operates.

## Mapping OIDC token claims to Fulcio OIDs

> | GitHub [(docs)][github-oidc-doc] | GitLab | CircleCI | Buildkite | Fulcio Certificate Extension | Why / Notes / Questions                                                                                                                                                                |

Author: For ref, current workload OIDC claims:
- Buildkite: https://github.com/sigstore/fulcio/pull/890#discussion_r1027648102
- Circle: https://circleci.com/docs/openid-connect-tokens/#format-of-the-openid-connect-id-token
- Gitlab: https://docs.gitlab.com/ee/ci/cloud_services/index.html#how-it-works
- GitHub (outdated): https://docs.github.com/en/actions/deployment/security-hardening-your-deployments/about-security-hardening-with-openid-connect#understanding-the-oidc-token

This specifies the username identity in the OtherName Subject Alternative Name, as
defined by [RFC5280 4.2.1.6](https://datatracker.ietf.org/doc/html/rfc5280#section-4.2.1.6).

> ### 1.3.6.1.4.1.57264.1.8 | Source URI

Reviewer 6: In order to disambiguate between "Source URI" and "Source Repository URI", how about:

```suggestion
### 1.3.6.1.4.1.57264.1.8 | Source Repository URI
```

Should include a fully qualified source repository URL that the build was based

> on.

Reviewer 6: ```suggestion
on, e.g. `https://github.com/sigstore/fulcio`.
```

> ### 1.3.6.1.4.1.57264.1.9 | Source Digest

Reviewer 6: ```suggestion
### 1.3.6.1.4.1.57264.1.9 | Source Repository Digest
```

An immutable reference to a specific version of the source code that the build

> was based upon. Should include the digest type followed by the digest, e.g.
> `sha1:abc123`.

Reviewer 6: ```suggestion
was based upon. Should include the digest type followed by the digest, e.g.
`sha1:abc123` for a git commit.
```

> ### 1.3.6.1.4.1.57264.1.10 | Source Ref

Reviewer 6: ```suggestion
### 1.3.6.1.4.1.57264.1.10 | Source Repository Ref
```

The source ref that the build run was based upon. For example: `refs/head/main`.

> ### 1.3.6.1.4.1.57264.1.11 | Source Identifier

Reviewer 6: ```suggestion
### 1.3.6.1.4.1.57264.1.11 | Source Repository Identifier
```

> The stable identifier for the source repository the workflow was based upon.

Reviewer 6: ```suggestion
The immutable identifier for the source repository the workflow was based upon, for e.g. `1234` if using a primary key.
```

### 1.3.6.1.4.1.57264.1.12 | Source Owner URI

> Fully qualified URL for the owner of the source repository that the build was based

Reviewer 7: nit: Adding an example for each of these fields would be useful.

Author: Good callout! @phillmv helpfully provided a bunch of examples. I'll add a terms section.

> on.

Reviewer 6: ```suggestion
on, e.g. `https://github.com/sigstore`
```

### 1.3.6.1.4.1.57264.1.13 | Source Owner Identifier

> The stable identifier for the owner of the source repository that the workflow was based upon.

Reviewer 6: ```suggestion
The immutable identifier for the owner of the source repository that the workflow was based upon, e.g. `5678` if using a primary key.
```

### 1.3.6.1.4.1.57264.1.14 | Build Config URI

A reference to the initiating build instructions.

### 1.3.6.1.4.1.57264.1.15 | Build Config Digest

An immutable reference to the specific version of the top-level build
instructions. Should include the digest type followed by the digest, e.g.
`sha1:abc123`.

### 1.3.6.1.4.1.57264.1.16 | Build Signer URI

Reference to specific build instructions that are responsible for signing. Can be the same as Build Config URI. For example a reusable workflow in GitHub Actions or a Circle CI Orbs.

> Note this value would also be set in the Subject Alternative Name (SAN).

Reviewer 6: For simplicity sake, let's just cut this line out. Even if it is the SAN, this value should still be filled out.

```suggestion
```

### 1.3.6.1.4.1.57264.1.17 | Build Signer Digest

An immutable reference to the specific version of the build instructions that is responsible for signing. Should include the digest type followed by the digest, e.g. `sha1:abc123`.

### 1.3.6.1.4.1.57264.1.18 | Build Trigger

The event or action that triggered the build.

### 1.3.6.1.4.1.57264.1.19 | Runner Environment

For platforms to specify whether the build took place in platform-hosted cloud infrastructure or customer-hosted infrastructure. For example: `platform-hosted` and `self-hosted`.

> 

Reviewer 6: ```suggestion

### 1.3.6.1.4.1.57264.1.20 | Runner Execution Identifier

An immutable value that can uniquely identify the build' sexecution. For example: `4043059413`

```

Reviewer 8: > but in the meantime being able to provide a non-falsifiable audit trail back to the specific build execution that requested the cert feels like an interesting/important property – especially in forensic contexts.

Would this one that is deprecated eventually? 

I understand the benefit of one, but this seems like a temporary solution to the broader problem of actually creating a **non-falsifiable builder** to get that audit trail

Reviewer 9: I think it actually makes sense. I would think of this kind of similar to the serial number for a cert. It adds a trust relation to it's origin (i.e its not at all intended for any policy decision).

Today, if we would like to understand if certificate from Fulcio is fraudulent or not, we have very limited information to go on. We have the the time it was created, but that's about it, the rest of the correlation would be a bit fuzzy. If we have the e.g. the run id, we know it's origin and can easier compare logs.

Having transparency is of course what enables this, similar to how it's done with WebPKI. What I think can be a problem here, is that the order of certificates created is on a different scale for signing certificates compared to domain certificates, so a "weak mapping" is harder to implement, and so including the run id would create a stronger link that's easier to automate on.

Reviewer 8: > I would think of this kind of similar to the serial number for a cert

Yeah, I was thinking about this, and wondering if it's something like a "session id" for the dex/email case.

My one question is: does it uniquely identify the identity/trust of the run? From what I understand, a single run_id can reference maybe many reusable workflows, which have different "identities". (Or is it the combo of the run_id and the job_workflow_ref/sha that can give you a unique reference and auditable issuance to the signer?)

What do you think then is the criticality of this extension? Honestly now maybe I can't decide if it's a MUST or very optional. MUST because it's an audit number to the OIDC token (does any other field suffice?), or optional because it's not related really to a useful part of the "identity" of the signer. Maybe I don't understand this: are you trying to audit the OIDC token issuance, or the invocation it was used in? The latter is definitely not "identity" policy related, so would be optional.

Reviewer 9: Yes, you are correct that we need to figure out how to make it unique, like a combination of`run_id/number/attempt`. This information together with e.g. the workflow_ref or sha as you propose. Not sure if the extension need to be cater for this, or if the "backing system" should be responsible for putting them altogether, I think that may be more useful as it's a bit more flexible/future-proof?

I would say it's RECOMMENDED? As from a general signer/identity perspective, it doesn't really add much?

Reviewer 9: A potential use-case I can see is for a CI provider to monitor the CT log and correlate each certificate with a workflow run to scan for forged certificates.

Reviewer 8: > I would say it's RECOMMENDED? As from a general signer/identity perspective, it doesn't really add much?

I'd be happy with that! It actually works for almost all extensions as well "here's the description of the extension and here's the "usage hint"" (which can help distinguish between critical for identity verification and informational and for what). WDYT @haydentherapper? 

Ideally the critical for identity verification line up with the required extensions Fulcio provides.

Other things, we should document and be clear what it's purpose is -- so that if some client abuses usage, it's up to them really but at least the intent for the extension is clear.

Author: I like this addition! As mentioned, this needs to include the run id and attempt to make it unique in GHA. Thoughts on following the lead of SLSA 1.0 draft and constructing a URI for this? https://github.com/slsa-framework/slsa/blob/main/docs/github-actions-workflow/v0.1/index.md#metadata

Something like `Run Invocation URI: github.server_url + "/actions/runs/" + github.run_id + "/attempts/" + github.run_attempt`. This way the value is immediately parseable and useful without having to know some custom format for GHA, e.g. `github.run_id + '-' + github.run_attempt`.

Author: `Build Invocation URI`?

Reviewer 9: While I don't think strict alignment with SLSA is relevant for Fulcio, I do think it's valuable to rely on prior art, so I like the idea or reusing this format.

Author: Added this as `Run Invocation URI` in the current proposal.

## 1.3.6.1.4.1.57264.2 | Policy OID for Sigstore Timestamp Authority

Not used by Fulcio. This specifies the policy OID for the [timestamp authority](https://github.com/sigstore/timestamp-authority)
that Sigstore operates.

> ## Mapping OIDC token claims to Fulcio OIDs

Reviewer 8: A proposed new claim: a version claim. I like the idea of sticking direct OIDC claims into the extension fields in the table. What I'm worried about is that the OID extensions are not versioned, and there is no way of allowing an OIDC provider to "update" the mapping (because we don't want to keep adding fields from this standardization). Having a Fulcio-integration "version" will help allow providers to update the table claims.

For e.g. if GH decides to add a sha prefix to the source digest, assuming that Fulcio is not sanitizing or modifying claims into this table, then that would be breaking to policy verifiers.

Reviewer 10: Agreed, and IMO this is another good motivation for a single "composite" extension that wraps all of these claims as suggested in #981 (although that will need to happen with a breaking version, not here).

Reviewer: Thought a bit more on this - Does a version that Fulcio maintains solve the problem of breaking policy verifiers? And furthermore, is that Fulcio's problem?

GH making a change to the format of the source digest would affect a verifier if they were directly consuming the OIDC token. This seems more like a problem for the OIDC providers, that they should version their tokens and we can then also include that version value in an extension. Fulcio isn't opinionated on each claim structure (at least that's what we decided after chatting - If it was opinionated, then we would need a version that fulcio maintains)

Adding a claim in a new extension should not be a breaking change, and clients should search for it and gracefully handle it not being present (verification policies can require the extension if it should be critical).

Reviewer: Chatting with Asra, throwing around some ideas (sorry if this is not clear, just trying to get this down to remember this):

* Each extension becomes a list of structs of {version, value}. If we have to update a extension value, we add a new struct to the list, with an updated version, while maintaining the previous version-value. 
* We must always operate in append-only, else we break verification for clients
* If we were to do something like #981, while being append-only, we'd have to duplicate the entire composite every time we'd do an update for any extension value for any CI provider
* The first bullet point here minimizes the duplications. 
* Another idea is a version-per-issuer, but this needs more thought

Reviewer 8: Thinking more on this: I think we should not complicate client logic with versioning logic, and push back on integrators who will be making version breaking changes to their  OIDC tokens. In a future world, I'd like OIDC tokens maybe to be versioned and for that to be represented in a claim, but I think Fulcio self-managing a versioning scheme would be pretty tough as an intermediate versioning layer between the OIDC provider and the client.

As in: let's push back on integrators who won't satisfy stable claims for the ones they are committing to tin the mapping.

Reviewer: Leaving this open in case anyone has any other thoughts, but I think we're all in agreement that we won't attempt to do any versioning.

| sha                | ??     | ??       | build_commit | Source Digest                | An immutable reference to a specific version of the source code. Should include the digest type followed by the digest, e.g. `sha1:abc123`.                                            |
| ref                | ref     | ??       | build_branch | Source Ref                   | The source ref that the build run was based upon. For example: refs/head/main.                                                                                                         |
| repository_id         | project_id     | ??       | ??        | Source Identifier     | Stable identifier for the owner of the source repository.                                                                                                                                      |
| repository_owner         | namespace_path     | ??       | ??        | Source Owner URI     | Fully qualified URL for the owner of the source repository.                                                                                                                                     |

> | repository_owner_id         | namespace_id     | ??       | ??        | Source Owner Identifier     | Stable identifier for the owner of the source repository.                                                                                                                                   |

Author: Q: For providers like Buildkite and Circle who don't also host the source, should they forward the underlying source/source owner ids in their workload identity? How do we track the "build identity" that executed the build and distinguish this from the source identity? Is the build config URI enough?

Reviewer 10: Someone more familiar with those CI services can probably answer better than me, but my expectation is that `repository_id` would be the be the "project" or "pipeline" ID on Buildkite or Circle, while `repository_owner_id` would be the ID of the `User` or other owner model that configured the pipeline.

That's contingent on these services having some kind of underlying relational design that looks roughly like the following:

```
User(id, ...) from table of USERS
Project(id, user_id_roles) from table of PROJECTS
```

...but I think that's probably a reasonable assumption 🙂

Author: Yeah would be good to get some clarity here. I was imaging `Source *` extensions would all point to the source repository. So if you use Circle to build with code on GitHub, Source URI would point to the github repo. Ref, digest would also point to the repo.

The reason we would want Source Owner Identifier and Source Identifier would be to spot transfers/resurrected repositories, in this example this would be the GitHub repo. Given this, I think we actually want Circle et al to provide these source/source owner ids in their OIDC claims. 

Build Config URI would presumably, in most cases, point to a build config file in the same repo. 

The Builder Signer URI, could in Circle's case be a [Circle Orb](https://circleci.com/orbs/) if one is used, otherwise this would point to the same above build config uri if one was not used.

Given this, there's actually no space in the current extensions for any build system identifiers like "project" or "pipeline" ID on Buildkite or Circle 🤔 

Question: do we need something specific here to track separate build system/build owner identifiers when this differs from the source system?

cc @haydentherapper what are your thoughts on above?

Reviewer: We should have a way to differentiate CI platforms, that should be a part of what’s validated by the verifier. Otherwise I could imagine a threat vector where an attacker creates a project on a different platform matching all of the project/repo details, but changing the source.

Author: > We should have a way to differentiate CI platforms, that should be a part of what’s validated by the verifier.

This makes sense, is this not already covered by the OIDC issuer extension?

Reviewer: Yes this is! OIDC issuer is sufficient

Reviewer 11: > but my expectation is that `repository_id` would be the be the "project" or "pipeline" ID on Buildkite or Circle, while `repository_owner_id` would be the ID of the `User` or other owner model that configured the pipeline.

I am familiar with Buildkite. Reading this spec, I would assume any OID mentioning "Source" or "Repository" would be some pointer back to, or meta data about the GitHub (in my case) repository and commit/ref that triggered the build.

And I'd expect `Run Invocation URI` to be something like `https://buildkite.com/mycompany/mypipeline/builds/118`

And my issuer (Buildkite) would be trusted to set the appropriate fields related to the source repository.

| workflow_sha       | ??     | ??       | ??        | Build Config Digest          | An immutable reference to the specific version of the top-level build instructions. Should include the digest type followed by the digest, e.g. `sha1:abc123`.                         |
| job_workflow_ref   | ??     | ??       | ??        | Build Signer URI             | Reference to specific build instructions that are responsible for signing. Can be the same as Build Config URI. For example a reusable workflow in GitHub Actions or a Circle CI Orbs. |
| job_workflow_sha   | ??     | ??       | ??        | Build Signer Digest          | An immutable reference to the specific version of the build instructions that is responsible for signing. Should include the digest type followed by the digest, e.g. `sha1:abc123`.   |
| event_name         | pipeline_source | ??       | ??        | Build Trigger                | The event or action that triggered the build.                                                                                                                                          |

> | runner_environment | ??     | ??       | ??        | Runner Environment           | For platforms to specify whether the build took place in platform-hosted cloud infrastructure or customer-hosted infrastructure. For example: `platform-hosted` and `self-hosted`.     |

Reviewer 6: ```suggestion
| runner_environment | ??     | ??       | ??        | Runner Environment           | For platforms to specify whether the build took place in platform-hosted cloud infrastructure or customer-hosted infrastructure. For example: `platform-hosted` and `self-hosted`.     |
| run_id | ?? | ?? | ?? | Runner Execution Identifier | An immutable value that can uniquely identify the build's execution. For example: `4043059413`
```

| iss                | ??     | ??       | ??        | Issuer                       | This already exists. For example: https://token.actions.githubusercontent.com                                                                                                          |
| exp                | ??     | ??       | ??        | N/A                          | Only used to validate the JWT.                                                                                                                                                         |
| nbf                | ??     | ??       | ??        | N/A                          | Only used to validate the JWT.                                                                                                                                                         |
| iat                | ??     | ??       | ??        | N/A                          | Only used to validate the JWT.                                                                                                                                                         |

> | repository         | ??     | ??       | ??        | Source URI                   | Should include a fully qualified repository URL.                                                                                                                                       |

Reviewer 5: I think right now GitHub's token has repository name here. I wonder of for backward compatibility it's worth to use `repository_url`?

Author: I don't think this claim exists atm for Actions. This probably means we would need to construct the source/repo URI in Fulcio based on the token issuer. Something like this for GHA:

```js
if (token.iss === 'https://token.actions.githubusercontent.com') {
  source_uri = `https://github.com/${token.repository}`
}
```

This specifies the username identity in the OtherName Subject Alternative Name, as
defined by [RFC5280 4.2.1.6](https://datatracker.ietf.org/doc/html/rfc5280#section-4.2.1.6).

### 1.3.6.1.4.1.57264.1.8 | Source Repository URI

> Should include a fully qualified source repository URL that the build was based

Reviewer: This one says "should" but we don't include should in other extensions. Is this referring to that the URL SHOULD be fully qualified, but may not be? Suggestion to rephrase: "Source repository URL that the build was based on. SHOULD be fully qualified. For example: blah", and then do the same with other URL extensions.

If being fully qualified is a MUST, we should note this since we'll need to enforce that.

Author: > Is this referring to that the URL SHOULD be fully qualified, but may not be?

👍 good spot, updated all the examples to make this clearer.

> on. For example: `example/repository`.

Reviewer 9: The comment mentions a fully qualified source repository URL, which according to the terminology is: "Complete URL with protocol", this example does not match that.

### 1.3.6.1.4.1.57264.1.9 | Source Repository Digest

An immutable reference to a specific version of the source code that the build
was based upon. For example: `abc123` git commit SHA.

### 1.3.6.1.4.1.57264.1.10 | Source Repository Ref

The source ref that the build run was based upon. For example: `refs/head/main`.

### 1.3.6.1.4.1.57264.1.11 | Source Repository Identifier

The immutable identifier for the source repository the workflow was based upon. For example: `1234` if using a primary key.

### 1.3.6.1.4.1.57264.1.12 | Source Repository Owner URI

Fully qualified URL for the owner of the source repository that the build was based

> on, e.g. `example`

Reviewer 9: Same here, the example is not fully qualified.

### 1.3.6.1.4.1.57264.1.13 | Source Repository Owner Identifier

The immutable identifier for the owner of the source repository that the workflow was based upon. For example: `5678` if using a primary key.

> ### 1.3.6.1.4.1.57264.1.14 | Build Config URI

Reviewer: To confirm, should this include the server URL? For GH, I see that workflow_ref and job_worflow_ref don't. If so, we should add SHOULD or MUST.

Author: Updated these to clarify the url SHOULD be fully qualified.

Just to nail down the GHA details, we're now prefixing these:
- `Run Invocation URI`: `server_url + repository + "/actions/runs/" + run_id + "/attempts/" + run_attempt`
- `Build Signer URI`:  `server_url + job_workflow_ref`

But not these:
- `Source Repository URI`: `repository` (owner/repo)
- `Source Repository Owner URI`: `repository_owner` (owner)
- `Build Config URI`: `workflow_ref`

.. should we be consistent across the board and prefix those with `server_url`? 😅

Author: TBH seems ok as is. Should we get this in as is and noodle on specific GHA details when implementing if needed?

Reviewer: I'm fine updating in a follow up, but my plan would be to use exactly what is in this doc when implementing for any CI provider (also makes it easier to audit). So if you want to prefix a URI with server_url, we should specify it.

A reference to the initiating build instructions. For example: `https://example.com/owner/repository/build-config.yml`.

### 1.3.6.1.4.1.57264.1.15 | Build Config Digest

An immutable reference to the specific version of the top-level build
instructions. For example: `abc123` git commit SHA.

> ### 1.3.6.1.4.1.57264.1.16 | Build Signer URI

Reviewer: Was there a previous iteration of this doc that called out that this would be the Subject Alternative Name? Do we want to include that?

> Reference to specific build instructions that are responsible for signing. Can be the same as Build Config URI. Could be a reusable workflow ref in GitHub Actions or a Circle CI Orb name. For example: `slsa-framework/slsa-github-generator/.github/workflows/generator_container_slsa3.yml@v1.4.0`.

Reviewer 9: Same here on fully qualified.

### 1.3.6.1.4.1.57264.1.17 | Build Signer Digest

> An immutable reference to the specific version of the build instructions that is responsible for signing. For example: `abc123` git commit SHA.

Author: @haydentherapper @asraa @kommendorkapten It just occurred to me that we won't just be dealing with Git SHA1's fairly soon. Circle has the concept of [Orb's](https://circleci.com/orbs/) that could be used to create a trusted builder and AFAICT these are OCI bundles so probs use SHA256.

Do you see any issue with not including the algo here if it's mostly SHA1 for all Git SCM's but SHA256 for Circle Orb's? It seems ok to push this on clients to figure out or by prior knowledge based on issuer 🤷 

From our chats I'm not sold on doing anything clever in Fulcio as this seems like it might just cause more problems down the road but keen to get your thoughts.

Reviewer 9: @feelepxyz I think keeping the value in Fulcio as-is from the integrator is preferred. I wouldn't think it's a big problem to omit the algorithm, as you said the clients should have prior knowledge on what to expect. This I think is also consistent with the other extensions in the certificate which values are specific per integration.

Reviewer: I agree with keeping the value as-is, though I'll play devils advocate here because I do see your concern. Are there any other extensions where verifiers are going to need to be aware of which CI provider the certificate came from? If not, then this is additional work for verifiers for just a single claim.
On the flip side, how will fulcio determine the algorithm? If we're just going to check digest length for example, that's something a client could do.

Reviewer: I don't see any other claims where it's expected to know the CI provider.

Slightly tangential, is hash_alg:digest standardized anywhere?

Reviewer 9: > If we're just going to check digest length for example, that's something a client could do.

Totally agree. A better option IMHO is to push for the CI vendors to switch to alg:digest. 

Also thinking how a verifier should consume the digest during policy execution. By definition the digest does not have any semantic structure, and so what a client can do is to keep an allowlist of known "good" digests for the builders. That would mean that the digest is an opaque string used to compare a against a list of approved values. In that case, the inclusion of algorithm would not make a difference. 

If we compare this with the subject digest in an in-toto attestation, the hash is important, as we are mapping artifacts (where the client needs to first compute the message digest) to attestations, which will likely be of high cardinality. For builders I would expect low cardinality and a highly specialized scenrio with much tighter policies, where the client typically don't compute a digest.

> Slightly tangential, is hash_alg:digest standardized anywhere?

I'm only aware of the OCI image spec: https://github.com/opencontainers/image-spec/blob/v1.0.1/descriptor.md#digests

Author: > On the flip side, how will fulcio determine the algorithm? If we're just going to check digest length for example, that's something a client could do.

Yeah was thinking either check length or just hardcode expected algo for a given issuer and claim but this might break in future. But doesn't feel like a major value add?

> I don't see any other claims where it's expected to know the CI provider.

You kinda need to know what to expect for each CI provider for all of them to be able to verify things. For `Runner Environment` this might be set to `github-hosted` and `circle-hosted` etc. 

I think what @kommendorkapten is is right, if the verifier is mainly checking extension values against approved values, none of this would be a problem as you're just comparing strings.

> Totally agree. A better option IMHO is to push for the CI vendors to switch to alg:digest.

Yup something needs to happen here across the industry to support sha256 in git 🙈

Reviewer: Closing on this, I believe we're all in agreement now to not include a hash algorithm and copy the value as-is, correct? We can always update during implementation if we see a need for it.

### 1.3.6.1.4.1.57264.1.18 | Build Trigger

The event or action that triggered the build. For example: `push`.

### 1.3.6.1.4.1.57264.1.19 | Run Invocation URI

> An immutable identifier that can uniquely identify the build execution. For GitHub Actions this needs to include both the run id and attempt. For example: `https://github.com/example/repository/actions/runs/1536140711/attempts/1`.

Reviewer: Should we omit "For GHA this needs to include ...", since that is documented below already?

| ref                | ref     | ??       | build_branch | Source Repository Ref                   | The source ref that the build run was based upon. For example: refs/head/main.                                                                                                         |
| repository_id         | project_id     | ??       | ??        | Source Repository Identifier     | Stable identifier for the owner of the source repository.                                                                                                                                      |
| repository_owner         | namespace_path     | ??       | ??        | Source Repository Owner URI     | Fully qualified URL for the owner of the source repository.                                                                                                                                     |
| repository_owner_id         | namespace_id     | ??       | ??        | Source Repository Owner Identifier     | Stable identifier for the owner of the source repository.                                                                                                                                   |

> | workflow_ref       | ??     | ??       | ??        | Build Config URI             | A reference to the initiating build instructions.                                                                                                                                      |

Reviewer: For this and job_workflow_ref, do we need to append `server_url` to make them into URIs?

Author: Yep 👍

This specifies the username identity in the OtherName Subject Alternative Name, as
defined by [RFC5280 4.2.1.6](https://datatracker.ietf.org/doc/html/rfc5280#section-4.2.1.6).

### 1.3.6.1.4.1.57264.1.8 | Source Repository URI

Should include a fully qualified source repository URL that the build was based
on. For example: `https://github.com/sigstore/fulcio`.

### 1.3.6.1.4.1.57264.1.9 | Source Repository Digest

An immutable reference to a specific version of the source code that the build
was based upon. Should include the digest type followed by the digest. For example:

> `sha1:abc123` for a git commit.

Reviewer 10: Not to make things more difficult here (sorry!), but is there any possibility we could make this more normative?

For example, we could enforce the following:

1. Only certain digest types are valid (e.g. `sha1`, `sha256`)
2. The digest itself must be a hexadecimal string, not another representation (like base64)
3. The digest itself must be a full digest, not a truncated one

(This is a good argument IMO for using ASN.1 structures, since we'd get a lot of these restrictions for "free" with ASN.1 definitions. But that's further out, so just listing these as plain language constraints would be good IMO.)

Reviewer: Our thought was that Fulcio should not be opinionated on claim values and it's up to the verifier to understand what claims will look like (i.e. they must consult the documentation of the provider, not Fulcio).

Reviewer 10: That makes sense! I'm a big fan of ratcheting down formats/flexibility wherever it can be ratcheted, but I can understand if this is a place where provider variability/diversity would make doing so too complicated or onerous.

Given that, I think many of these claims would benefit from `MAY` language that emphasize that providers may choose to fill these values as they please, e.g.:

```
Providers MAY choose to emit this extension's value in other formats, and consumers MUST NOT assume the format above. 
```

Reviewer: +1, I think we should be more clear in describing examples as optional suggestions.

We wanted to make it as easy as possible to onboard new providers, which was another justification for fulcio just forwarding provided claims. Otherwise we're requiring CI platforms to conform to fulcio's requirements, which would likely not happen, or fulcio would mutate the claim before certifying it, which increases trust in fulcio.

Reviewer 10: > Otherwise we're requiring CI platforms to conform to fulcio's requirements, which would likely not happen, or fulcio would mutate the claim before certifying it, which increases trust in fulcio.

Oh, this is a great point -- I hadn't considered how this places additional trust in Fulcio. Thanks!

Author: > Given that, I think many of these claims would benefit from `MAY` language that emphasize that providers may choose to fill these values as they please, e.g.:

@woodruffw good idea! I added this as top-level call out to cover all the values.
